core/trino-main/src/main/java/io/trino/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -1280,7 +1281,18 @@ private boolean joinDynamicFilteringExpression(Expression expression, Collection
                 }
                 comparison = (ComparisonExpression) expression;
             }
-            return comparison.getLeft() instanceof SymbolReference && comparison.getRight() instanceof SymbolReference;
+
+            return dynamicFilteringApplicable(comparison, leftSymbols, rightSymbols);
+        }
+
+        // sub expression that uses right source has to be a symbol (left can be an arbitrary expression)
+        private boolean dynamicFilteringApplicable(ComparisonExpression comparison, Collection<Symbol> leftSymbols, Collection<Symbol> rightSymbols)
+        {
+            // right is a symbol and there is NO MISALIGNMENT between comparison order and left/right source order
```

Comment:
Rather than this comment, change comment to something like:
```
Build side expression must be a symbol reference, since DynamicFilterSourceOperator can only collect column values (not expressions)
```